docs/proposals/api-group.md

```diff
@@ -0,0 +1,104 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/api-group.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Supporting multiple API groups
+
+## Goal
+
+1. Breaking the monolithic v1 API to modular groups
+
+2. Supporting different versions in different groups
+
+3. Supporting identically named kinds to exist in different groups
+
+4. Enabling/disabling groups individually
+
+5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
+
+6. Allowing users to omit group name when using kubectl
+
+
+## Bookkeeping for groups
+
+1. No changes to API spec:
+
+  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
```

Comment:
To clarify - I think "api version" in an object as an opaque string - a
vector that represents how to uniquely relate Kind and the object provided.
 "group" as we've described it is a convention that on the server side is
enforced, but on the client side can be opaque.  If I post a resource to
the server without apiVersion or kind to an endpoint that is versioned and
grouped, there is no ambiguity to me (the client author).

In our internal objects, if we are dealing with objects opaquely (rest
client and resource.Helper) the version has to be opaque anyway, and
ClientForMapping is the one who has to know about the version.  Clients
can't convert / process objects anyway, they can just say whether they know
where on a server to send "kind" for a given "apiVersion".

On Thu, Aug 20, 2015 at 3:19 PM, Clayton Coleman <ccoleman@redhat.com>
wrote:

> I want these values to be opaque to most client code.  The only thing that
> needs to know about group is a RESTMapper and the inner guts of conversion.
>
> On Thu, Aug 20, 2015 at 3:08 PM, Paul Morie <notifications@github.com>
> wrote:
>
>> In docs/proposals/api-group.md
>> <https://github.com/kubernetes/kubernetes/pull/12951#discussion_r37568922>
>> :
>>
>> > +2. Supporting different versions in different groups
>> > +
>> > +3. Supporting identically named kinds to exist in different groups
>> > +
>> > +4. Enabling/disabling groups individually
>> > +
>> > +5. Laying the basis for [API Plugin](../../docs/design/extending-api.md)
>> > +
>> > +6. Allowing users to omit group name when using kubectl
>> > +
>> > +
>> > +## Bookkeeping for groups
>> > +
>> > +1. No changes to API spec:
>> > +
>> > +  We will not add new field in the API spec to hold the group information. Group will be part of the TypeMeta.APIVersion field, in the form of `group/version`.
>>
>> +1 to this, we will wind up with some utility method that parses this out
>> into group and version that we call all over the place if we don't add a
>> distinct field.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/12951/files#r37568922>.
>>
>
>
>
> --
> Clayton Coleman | Lead Engineer, OpenShift
>



-- 
Clayton Coleman | Lead Engineer, OpenShift